AppBundles/PluginUtilities/ParametersExtractor.cs

```diff
@@ -0,0 +1,172 @@
+// restricted rights notice below appear in all supporting
+// documentation.
+//
+// AUTODESK PROVIDES THIS PROGRAM "AS IS" AND WITH ALL FAULTS.
+// AUTODESK SPECIFICALLY DISCLAIMS ANY IMPLIED WARRANTY OF
+// MERCHANTABILITY OR FITNESS FOR A PARTICULAR USE.  AUTODESK, INC.
+// DOES NOT WARRANT THAT THE OPERATION OF THE PROGRAM WILL BE
+// UNINTERRUPTED OR ERROR FREE.
+/////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Inventor;
+using Newtonsoft.Json;
+using Shared;
+
+namespace PluginUtilities
+{
+    public class ParametersExtractor
+    {
+        public void Extract(Document doc, Parameters parameters, InventorParameters updateResultParameters = null)
+        {
+            // extract user parameters
+            InventorParameters allParams = ExtractParameters(doc, parameters);
+
+            // save current state
+            LogTrace("Updating");
+            doc.Update2();
+            LogTrace("Saving");
+            doc.Save2(SaveDependents: true);
+
+            // detect iLogic forms
+            iLogicFormsReader reader = new iLogicFormsReader(doc, allParams);
+            iLogicForm[] forms = reader.Extract();
+            LogTrace($"Found {forms.Length} iLogic forms");
+            foreach (var form in forms)
+            {
+                LogTrace($" - {form.Name}");
+            }
+
+            // Choose set of parameters to use with the following algorithm:
+            // - extract all iLogic forms from the document
+            //   - keep only 'user parameters' from a form
+            // - use _first_ iLogic form with non-empty list of parameters
+            // - if no forms - use UserParameters from the document
+            InventorParameters resultingParameters;
+            var candidate = forms.FirstOrDefault(form => form.Parameters.Count > 0);
+            if (candidate != null)
+            {
+                LogTrace($"Using '{candidate.Name}' form as a parameter filter");
+                resultingParameters = candidate.Parameters;
+            }
+            else
+            {
+                LogTrace("No non-empty iLogic forms found. Using all user parameters.");
+                resultingParameters = ExtractParameters(doc, parameters.UserParameters);
+            }
+
+            // if parameters from an update result were passed in, if validation errors were set on any replace values with them
+            if (updateResultParameters != null)
+            {
+                foreach (KeyValuePair<string, InventorParameter> updateResultParam in updateResultParameters)
+                {
+                    if (updateResultParam.Value.ErrorMessage != null)
+                    {
+                        resultingParameters[updateResultParam.Key] = updateResultParam.Value;
+                    }
+                }
+            }
```

Comment:
This is mostly just moved from the ExtractParameters plugin except this block is added to support the update workflow.